### ISSUE TITLE
x15 pumps - incorrect "bolus many not have succeeded" notification after bolus

### ISSUE THREAD

**Issue by dabear** [LoopKit/Loop#587]
I am using a medtronic 715wws pump.
I am using offline-mode only with dexcom share.

Loop 1.4 no problems.
Loop 1.5: boluses fail every time with the attached error message
![img_0061](https://user-images.githubusercontent.com/442324/30751548-e4a728a4-9fb9-11e7-91b9-bf1185fc059b.PNG)

The pump actually successfully enacts the bolus every time, despite this error message. 20 minutes after, the loop sends a notification that loop hasn't run successfully in 20 minutes and the loop circle is red.

Here's the output from xcode terminal while running a bolus:
https://gist.github.com/dabear/eab8ac4b982307e44167fc97c3a6f9a6

**Comment by lostinthemidwest** [comment]
I have the same problem with the 715. I have a rfcat; I'd be happy to send you capture logs if you let me know how to use it and what to capture.

**Comment by dabear** [comment]
@lostinthemidwest this is a regression in Loop 1.5 it seems. If you take a look at the gist I provided earlier, the pump is indeed sending a for of ACK (an "OK, SUCCESS" message) to Loop. For some reason there are probably multiple types of ACKs and loop doesn't know about the specific ACK sent from our pumps.

**Comment by dabear** [comment]
I guess this is relevant: https://github.com/ps2/rileylink_ios/commit/d776d79ccb1280b123d144f41bb39d2ecc39f8af

**Comment by dabear** [comment]
This hack worked for my pump. I can now bolus again from loop: https://github.com/dabear/rileylink_ios/commit/02a5a3cd89d3932670bc4074e3dcc0344a0ad4bc

**Comment by ps2** [comment]
I'm wondering if there is some firmware version difference going on here.  I would appreciate if any x15 pump users are able to try out 1.5 and report here their firmware version, and whether or not 1.5 works for them.

**Comment by ps2** [comment]
I believe the firmware version on the pump I tested this code with is `1.2A 1.1 0B 0B`

**Comment by dabear** [comment]
Paradigm 715ww
Ver 2.3A 1.1 0B 0B

**Comment by christerjensen** [comment]
Dabear's solution worked. 715 ww 2.4A 1.1

**Comment by dabear** [comment]
For anyone doing this manual update using my hack, please note the following:

You must update carthage to latest version.
You will need to update your Loop's carthage file to point to your own rileylink_ios repository

It is really, really important that you remove the cached version of rileylink_ios before running "carthage update"

```bash
rm -rf $HOME/Library/Caches/org.carthage.CarthageKit/dependencies/rileylink_ios/
```

If you get the error no tagged version for dexcom share client, delete the cached version before running "carthage update"
```
rm -rf $HOME/Library/Caches/org.carthage.CarthageKit/dependencies/dexcom-share-client-swift/
```

**Comment by john36m** [comment]
My backup pump is a 515.  I hope this gets fixed eventually.

**Comment by jlaes** [comment]
Our pump is a 515 WW, firmware is 2.3A and we are having the same issue, will try the hack tonight for it

**Comment by jlaes** [comment]
dabear is it enough just to follow your instructions on removing the cache and doing the changes in: https://github.com/dabear/rileylink_ios/commit/02a5a3cd89d3932670bc4074e3dcc0344a0ad4bc

Is there anything else that needs to be changed for it to work somewhere else? Do I need to change also this:

https://github.com/ps2/rileylink_ios/commit/d776d79ccb1280b123d144f41bb39d2ecc39f8af

 I'm guessing the hack could fix our 515 WW 2.3A which is having these issues.

I would love to try this but unsure what all I need to do to go through it.

**Comment by dabear** [comment]
you don't need to change the last part

**Comment by rodzilla55** [comment]
515 2.0A 1.1 0B 0B
Haven’t tried the fix yet.

**Comment by rodzilla55** [comment]
@dabear I am an experienced coder, but completely novice on the IOS platform and unfortunately pretty green on github. I'm not understanding how to make the code mod above (where to make it). I have only downloaded and deployed the loop app... and I see that it imports "RileyLinkKit" and "MinimedKit", but I don't know where those source files are to modify them (specifically looking for pumpmodel.swift)?

**Comment by dabear** [comment]
@rodzilla55 loop imports a binary version of rileylink_ios which needs to be edited. You will have to do two things;

1) Fork a copy of rileylink_ios and do the changes in rileylink_ios mentioned previously.
2) Update Loop to use your new version of rileylink_ios.
```
rm -rf $HOME/Library/Caches/org.carthage.CarthageKit/dependencies/rileylink_ios/
cd into_loop_dir
vim Cartfile

```
<img width="412" alt="cartfile" src="https://user-images.githubusercontent.com/442324/34905653-ba79fe74-f85d-11e7-8ddb-02800dd1a9d5.png">
Run "carthage update" to utilize your new version. Carthage will compile the new version of rileylink_ios for you and loop will use that version now. You should inspect "cartfile.resolved" after, to make sure that loop is actually using your version of rileylink_ios.

**Comment by dabear** [comment]
I have now temporarily access to another 715 pump, this time a US version
Ver 1.2A 1.1 0B 0B
This pump does not need the workaround described here.

@ps2 What do you think about making expectedResponseType an array of [.errorResponse,.pumpAck] for the x15 models (and an array with one entry of [.pumpAck] for all others?). Shouldn't that solve the problem? Will it have any negative impact?

**Comment by xpucuto** [comment]
Ok so updated my RL branch and got all the suggested changes in my repo. Haven't personally made the changes as @ps2 had already merged them.  changed the rileylink source in the Carthage file with my repo and did Carthage update. Then rebuilt Loop but I still get the error when sending a bolus ? 
I am using WW 715 VER 2.3A 1.1 0B 0B
Amy ideas what have I missed ?

**Comment by dabear** [comment]
@xpucuto the problem is that you merged your version from rileylink_ios dev branch. Your assumption that dev branch fixes the problem looks incorrect to me on first glance. In anycase, follow the receipe, use the master branch and apply the changes mentioned here. 

Remember to do this again:
```
rm -rf $HOME/Library/Caches/org.carthage.CarthageKit/dependencies/rileylink_ios/
```
then update your cartthage file, run carthage update and rebuild loop

**Comment by xpucuto** [comment]
@dabear I did use my master branch 
github "LoopKit/LoopKit" == 1.5.3
github "LoopKit/xDripG5" ~> 0.8.1
github "i-schuetz/SwiftCharts" == 0.6.1
github "mddub/dexcom-share-client-swift" == 0.4.1
github "mddub/G4ShareSpy" == 0.3.3
github "xpucuto/rileylink_ios" "master"
github "amplitude/Amplitude-iOS" ~> 3.8.5

**Comment by xpucuto** [comment]
Ok I so built 1.5.2 today , changed my RL repo in the Carthage file , did the two commands @dabear mentioned above before the Carthage update and still got the error when performing a  bolus . 
here is the info from Xcode 

2018-02-01 13:25:56.799167+0200 Loop[18868:26559923] Updated response count: 180
2018-02-01 13:25:57.190082+0200 Loop[18868:26559923] Response to command: f77ca969b1d15c9a68d5595638d68f18e5555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555b0d5
2018-02-01 13:25:57.190742+0200 Loop[18868:26562320] Pump responding despite our wake timer having expired. Extending timer
2018-02-01 13:25:57.191290+0200 Loop[18868:26562320] Writing command to data characteristic: 0500000000000000b403a969b1d15c9ab0e5559a3000
2018-02-01 13:25:57.608955+0200 Loop[18868:26559923] Updated response count: 181
2018-02-01 13:25:57.969992+0200 Loop[18868:26559923] Response to command: 187da969b1d15c9ab0e5635635555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555558ea5
2018-02-01 13:25:57.970596+0200 Loop[18868:26562320] Writing command to data characteristic: 0500000000000000b403a969b1d15c9ad32555ab4000
2018-02-01 13:25:58.303484+0200 Loop[18868:26559923] Updated response count: 182
2018-02-01 13:25:58.359445+0200 Loop[18868:26559923] Response to command: 107ea969b1d15c9a566555ca65
2018-02-01 13:25:58.359855+0200 Loop[18868:26562320] Writing command to data characteristic: 0500000000000000b403a969b1d15c9ad32571571555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555366000
2018-02-01 13:25:59.109845+0200 Loop[18868:26559923] Updated response count: 183
2018-02-01 13:25:59.170860+0200 Loop[18868:26559923] Response to command: 0f7fa969b1d15c9a566555ca65
2018-02-01 13:25:59.174961+0200 Loop[18868:26562320] [Bolus] (
        (
        "RileyLinkKit.SetBolusError.uncertain(RileyLinkKit.PumpCommsError.unexpectedResponse(PumpMessage(carelink, pumpAck, 614028, 00), from: PumpMessage(carelink, bolus, 614028, 0101000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000)))"
    )
)
2018-02-01 13:25:59.179675+0200 Loop[18868:26562320] ------------------------ Set normal bolus ---------------------------
2018-02-01 13:25:59.183845+0200 Loop[18868:26559923] idleDetectDispatchGroup empty

**Comment by christerjensen** [comment]
I also had problems getting the right rileylink_ios - have you check the rileylink_ios ID is the same as in your Github fork?

The chartage update should have a rileylink_ios number ("74ab2xxxxxxxxx") that matches your Rileylink_ios fork in Github.


https://scontent.ftrd1-1.fna.fbcdn.net/v/t35.0-12/22751489_10154825090687115_695949521_o.jpg?oh=d3039e2831d59404ed563d84fa2731de&oe=5A756ADD

https://scontent.ftrd1-1.fna.fbcdn.net/v/t35.0-12/22711671_10154959212826512_1118976936_o.png?oh=c7bda4a3ff5c7fc97dc16ca07f0e7501&oe=5A7572D8

**Comment by christerjensen** [comment]
@xpucuto You did change the PumpModel.swift code?

![image](https://user-images.githubusercontent.com/11073675/35678102-5c2ec292-0752-11e8-9ef1-1857dc288910.png)

**Comment by christerjensen** [comment]
Checked your repository @xpucuto - you forgot to change line 67.

This is from your pumpmodel.swift
![image](https://user-images.githubusercontent.com/11073675/35678261-f706b130-0752-11e8-8b58-4fc23988c56b.png)

**Comment by xpucuto** [comment]
Oh I am sorry . Guess I have missed that . Thank you for double checking it for me . Will fix it tonight ;)

**Comment by DConmy** [comment]
@ps2    

I have a 515LNAS with firmware 2.1A 1.1 0B0B that fails on boluses with the current code (rileylink_ios/master).    While I could do the workaround, I'd like to help get a cleaner change that can get merged into master.

Please let me know if there is data I can collect, or code changes I could make to help understand the difference between what you are seeing, and others that have failures that are seeing.

I see the change described in this thread is basically backing out the change you did in Aug.  I assume that was trying to fix something else, so backing it out, in general, would affect whatever it was trying to fix.

**Comment by Kdisimone** [comment]
I posted a merged version of Loop 1.5.3dev and the x15 fix listed above.  Just an fyi if anyone wants it.

https://github.com/Kdisimone/Loop/tree/x15-dev-fix

**Comment by jeremybarnum** [comment]
We are seeing this problem on a 515 (will post firmware shortly - not with the pump now) after moving to current master branch from 1.5.0.  Easy enough to fix by following above instructions but are others seeing this and any reason not to fix it in master?

**Comment by dabear** [comment]
The problem is that some x15 behaves differently. So if you fix it for some pumps, the others will fail.

**Comment by Kdisimone** [comment]
@diggabyte. I think we’ve known that for awhile because other earlier versions of Loop also used to have no issues for some x15 users. In order to figure out what’s happening, and IF pete (or other willing volunteer) had the time, he’d need both a no-error and error 515 to compare outputs from. I’ve found an error 515 and will be getting a non-error 515 soon...but I don’t expect he’d have priority time necessarily to look into this especially as there is a fairly easy work around.

**Comment by diggabyte** [comment]
@Kdisimone thanks. So if I understand this correctly, previously, it was "working" because `returnsErrorOnBolus` was true and thus expected response type was indeed an error, which was ignored. 

However since updating, the response is no longer an error (for whatever reason, TBD), which is unexpected. So the workaround is to modify the code so that we no longer expect an error for x15, correct?

**Comment by diggabyte** [comment]
Unfortunately, I've applied the patch and am still getting the error. So this workaround may not work in all cases. Here's the steps, just to illustrate I haven't missed anything:

- Forked rileylink_ios, [applied the patch](https://github.com/ps2/rileylink_ios/commit/d46f45ab666f44e4e70b5c4ef86b7ca7df5a29f6), cleared - updated Carthage to point to my fork, master branch
- clear carthage cache (`rm -rf ~/Library/Caches/carthage`)
- ran `carthage update`
- verified `Carthage.resolved` contains my fork's hash (`d46f45a...`)
- build, Loop, execute bolus

Expected: no error
Actual: Still produces the [unexpected response error](https://gist.github.com/diggabyte/f0f5399ce49e11705f0e6525b8d151f2)

**Comment by jeremybarnum** [comment]
@diggabyte have you managed to get it to work? I'm about to try, so would love to know if you've figured it out.

**Comment by Kdisimone** [comment]
It’s worked for me...

**Comment by diggabyte** [comment]
@jeremybarnum @Kdisimone yeah, tried several times and double-checked all my steps with `v2.3A 1.1`. Still produces the error. The strange part is that it never occurred with my builds via `xcode 9.3 / iOS 11.3`. It only surfaced when I updated to `xcode 9.4 / iOS 11.4`' and now I can't get it to go away.  

Are both on 9.4 / 11.4 as well?

**Comment by Kdisimone** [comment]
Yes. 9.4/11.4

**Comment by Kdisimone** [comment]
I’ll try building with your RL_ios today and see if it works for me.

**Comment by diggabyte** [comment]
@Kdisimone thank you. On my fork, I've patched both [master](https://github.com/diggabyte/rileylink_ios/tree/master) and also created a [515-patch branch](https://github.com/diggabyte/rileylink_ios/tree/515-patch) whose parent was the 2.0.0 tag.

**Comment by Kdisimone** [comment]
Which Loop branch are you trying?

**Comment by diggabyte** [comment]
I've been using Loop 1.5.6 (master), both when I was on 9.3/11.3 and also 9.4/11.4 xcode/ios.

**Comment by diggabyte** [comment]
I believe I finally got to the bottom of why my build was still showing the error despite the workaround being applied. 

It seems the carthage binaries were not actually being recompiled successfully because the following xcode  preference was never set:
`Xcode -> Preferences -> Locations -> Command Line Tools`

As a result, when running `carthage update` there was a hard-to-see error listed in the output: 
`xcrun: error: unable to find utility "xcodebuild", not a developer tool or in PATH` 

Some googling revealed that the `Command Line Tools` location must be set _from within xcode itself_. For whatever reason, this was never set by default. I had to manually set it, as seen from this screenshot from a stackoverflow user:

![image](https://user-images.githubusercontent.com/857573/42181531-61741622-7df8-11e8-8ee4-4fffc3c47103.png)

After setting this preference and again running `carthage update` I could see that the binaries were indeed being built from scratch now without any error messages. Previously, it was just using the existing binaries, thus the patch never seemed to work.

**Comment by mieslep** [comment]
We're on `v1.9.3-spike-master` (from `Kdisimone/Loop`) with a 515WWL v2.3A 1.1, and a completely fresh build (new-to-Loop this weekend!), and getting this error.  Trying to follow above and it seems this problem hasn't been resolved on `master` as yet?  So (being super-familiar with git but not in the least familiar with iOS development) trying to work out the most sensible way to try to get this patch into the main Loop `master` branch?

**Comment by Kdisimone** [comment]
The “solution” is is just a bandaid. Ideally a more comprehensive fix would be developed before landing in master. If you want to stay on my branch, I’d recommend using the x15-fix (or whatever it’s exactly named) branch I have and making the spike edits yourself on that one.

**Comment by mieslep** [comment]
Following https://github.com/LoopKit/Loop/issues/587#issuecomment-401095194 from @diggabyte , with rileylink_ios v2.0.7 and @Kdisimone 's `spike-master` branch (at v1.9.3 plus Spike-ey bits) and the error is "resolved".

**Comment by Katrin620** [comment]
Hi all -  it is amazing how much work has gone into assisting others with this issue. I have to admit though that I have been searching desperately for a description how to implement this fix 'for dummies'. I am completely lost with the descriptions. I can find and edit the lines, but doing something in github, forking and catharge - this sounds like greek. I've followed the step by step descriptions in the loopdocs, but this seems way off the deep end now...

**Comment by marionbarker** [comment]
I believe this was fixed in Loop v2.0 and thus can be closed.

**Comment by ps2** [comment]
Based on logs seen from a user using a 715 pump recently on dev, this issue was *not* resolved.  I had a new theory that involved our retry code, which I've changed here: https://github.com/ps2/rileylink_ios/commit/8bd5db161824112f3f7ba880b91a0266523a5520
